pkg/distributor/receiver/shim.go

```diff
@@ -0,0 +1,318 @@
+package receiver
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/go-logfmt/logfmt"
+	"github.com/grafana/dskit/services"
+	"github.com/opentracing/opentracing-go"
+	prom_client "github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/logging"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/configunmarshaler"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.opentelemetry.io/collector/receiver/otlpreceiver"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/trace"
+
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"github.com/grafana/loki/pkg/logproto"
+	util_log "github.com/grafana/loki/pkg/util/log"
+)
+
+const (
+	LogFormatJSON   = "json"
+	LogFormatLogfmt = "logfmt"
+)
+
+var (
+	metricPushDuration = promauto.NewHistogram(prom_client.HistogramOpts{
+		Namespace: "loki",
+		Name:      "distributor_push_duration_seconds",
+		Help:      "Records the amount of time to push a batch to the ingester.",
+		Buckets:   prom_client.DefBuckets,
+	})
+)
+
+type BatchPusher interface {
+	Push(ctx context.Context, req *logproto.PushRequest) (*logproto.PushResponse, error)
+}
+
+type receiversShim struct {
+	services.Service
+
+	receivers    []component.Receiver
+	pusher       BatchPusher
+	logger       log.Logger
+	format       string
+	drainTimeout time.Duration
+}
+
+func (r *receiversShim) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+// New
+// handler "/v1/logs" go.opentelemetry.io/collector/receiver/otlpreceiver/otlp.go:229
+func New(receiverCfg map[string]interface{}, pusher BatchPusher, receiverFormat string, receiverDrainTimeout time.Duration, middleware Middleware, logLevel logging.Level) (services.Service, error) {
```

Comment:
`logLevel` is never used. (also, afaik, we don't pass a log level anywhere)

```suggestion
func New(receiverCfg map[string]interface{}, pusher BatchPusher, receiverFormat string, receiverDrainTimeout time.Duration, middleware Middleware) (services.Service, error) {
```